### PR TITLE
fail if service name != service account name

### LINKED ISSUE #237: First container name might not match service account
When using connect inject with ACLs, if the first container's name doesn't match the service account name then the consul login will fail and the service won't be registered. We should investigate some better logging or error messaging in this case. Or we could use the service account name instead of the first container's name.

### PR DESCRIPTION
**Pull request by ashwin-venkatesh**
This check only occurs when ACLs are enable with connectInject.

Tested that it only performs this check when ACLs are enabled. Screenshot of error when incorrect pod config is applied

Closes #237 

Signed-off-by: Ashwin Venkatesh <ashwin@hashicorp.com>

Screen Shot 2020-06-19 at 10 54 48 AM

### CODE REVIEW

File: connect-inject/container_init.go [hunk 1]
```diff
@@ -81,6 +82,13 @@ func (h *Handler) containerInit(pod *corev1.Pod, k8sNamespace string) (corev1.Co
 		panic("No service found. This should be impossible since we default it.")
 	}
 
+	// If ACLs are enabled, the ACL token is assumed to be in the ServiceAccount that
+	// shares the name with the Service. If the name of the service does not match that
+	// of the ServiceAccount, the token will not be found leading to an opaque 403 error.
+	if data.AuthMethod != "" && data.ServiceName != pod.Spec.ServiceAccountName {
+		return corev1.Container{}, errors.New(fmt.Sprintf(`serviceAccountName "%s" does not match service name "%s"`, pod.Spec.ServiceAccountName, data.ServiceName))
```
Comment (lkysow) [inline connect-inject/container_init.go:89]:
```suggestion
		return corev1.Container{}, fmt.Errorf("serviceAccountName %q does not match service name %q", pod.Spec.ServiceAccountName, data.ServiceName)
```

I think this can be simplified. `%q` quotes the string.

File: connect-inject/container_init.go [hunk 2]
```diff
@@ -81,6 +82,13 @@ func (h *Handler) containerInit(pod *corev1.Pod, k8sNamespace string) (corev1.Co
 		panic("No service found. This should be impossible since we default it.")
 	}
 
+	// If ACLs are enabled, the ACL token is assumed to be in the ServiceAccount that
+	// shares the name with the Service. If the name of the service does not match that
+	// of the ServiceAccount, the token will not be found leading to an opaque 403 error.
```
Comment (lkysow) [inline connect-inject/container_init.go:87]: Can you reword this to say that the ACL token returned from `consul login` is only valid for a service with the same name as the serviceAccountName. I think the "assumed to be in the ServiceAccount" doesn't fully capture what's going on because the ACL token isn't in the service account. We exchange the service account token for a consul ACL token via `consul login`.

File: connect-inject/container_init_test.go [hunk 3]
```diff
@@ -1514,3 +1517,28 @@ func TestHandlerContainerInit_Resources(t *testing.T) {
...
+					Name: "serviceName",
+				},
+			},
+			ServiceAccountName: "notServiceName",
+		},
+	}
+
+	_, err := h.containerInit(pod, k8sNamespace)
+	require.EqualError(err, `serviceAccountName "notServiceName" does not match service name "foo"`)
+}
```
Comment (lkysow) [inline connect-inject/container_init_test.go:1544]: Can you add another test for when there is no annotation and we're using the first pod name and also for when ACLs are not enabled and so the names can be mismatched with no error.

Comment (ashwin-venkatesh) [reply connect-inject/container_init_test.go:1544]: I thought of writing that test but between https://github.com/hashicorp/consul-k8s/blob/7e9d6501f8431f9b1b7997b4a23341a4435758f7/connect-inject/container_init.go#L79:L83 and https://github.com/hashicorp/consul-k8s/blob/7e9d6501f8431f9b1b7997b4a23341a4435758f7/connect-inject/handler.go#L425:L434, the annotation for `serviceName` is always set and it will panic if the annotation is not.

Comment (lkysow) [reply connect-inject/container_init_test.go:1544]: Gotcha, makes sense!

Comment (lkysow) [reply connect-inject/container_init_test.go:1544]: Oh, what about a test for when ACLs are not enabled? i.e. `Handler.AuthMethod == ""`.

Comment (ashwin-venkatesh) [reply connect-inject/container_init_test.go:1544]: there are other tests up top that would fail and hence i didn’t write one explicitly. i can add one though.

Comment (lkysow) [reply connect-inject/container_init_test.go:1544]: Yeah I think an explicit one makes sense here. I looked at the other tests and they don't seem targeted enough to this use-case imo.

Comment (ashwin-venkatesh) [reply connect-inject/container_init_test.go:1544]: Done 💥
